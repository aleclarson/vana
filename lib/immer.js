"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var immer_1 = require("immer");
var common_1 = require("./common");
var commit_1 = require("./funcs/commit");
var symbols_1 = require("./symbols");
exports.immer = new immer_1.Immer({
    onAssign: function (state, prop) {
        state.assigned[prop] = true;
    },
    onDelete: function (state, prop) {
        state.assigned[prop] = false;
    },
    onCopy: function (state) {
        var base = state.base;
        var target = base && base[symbols_1.$O];
        if (target) {
            var copy = state.copy, assigned = state.assigned;
            if (common_1.isArray(base)) {
                // $O is missing, because Immer only copies the indices.
                common_1.definePrivate(copy, symbols_1.$O, target);
            }
            target._rebind(copy);
            for (var prop in assigned) {
                commit_1.commit(target, base[prop], copy[prop], prop, !assigned[prop], state);
            }
            commit_1.commit(target, base, copy, null, false, state);
        }
    },
});
// tslint:disable:no-default-export
exports.default = exports.immer.produce;
exports.produce = exports.immer.produce;
exports.setAutoFreeze = exports.immer.setAutoFreeze.bind(exports.immer);
exports.setUseProxies = exports.immer.setUseProxies.bind(exports.immer);
