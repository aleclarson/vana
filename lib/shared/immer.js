"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var immer_1 = require("immer");
var _1 = require(".");
var commit_1 = require("../core/commit");
exports.immer = new immer_1.Immer({
    onAssign: function (state, prop) {
        state.assigned[prop] = true;
    },
    onDelete: function (state, prop) {
        state.assigned[prop] = false;
    },
    onCopy: function (state) {
        var base = state.base;
        var target = base && base[_1.$O];
        if (target) {
            var copy = state.copy, assigned = state.assigned;
            if (_1.isArray(base)) {
                // $O is missing, because Immer only copies the indices.
                _1.definePrivate(copy, _1.$O, target);
            }
            target._rebind(copy);
            for (var prop in assigned) {
                commit_1.commit(target, base[prop], copy[prop], prop, !assigned[prop], state);
            }
            commit_1.commit(target, base, copy, null, false, state);
        }
    },
});
// tslint:disable:no-default-export
exports.default = exports.immer.produce;
exports.produce = exports.immer.produce;
exports.setAutoFreeze = exports.immer.setAutoFreeze.bind(exports.immer);
exports.setUseProxies = exports.immer.setUseProxies.bind(exports.immer);
