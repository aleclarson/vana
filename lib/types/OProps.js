"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// tslint:disable:variable-name
var immer_1 = require("immer");
var common_1 = require("../common");
var commit_1 = require("../funcs/commit");
var freeze_1 = require("../funcs/freeze");
var symbols_1 = require("../symbols");
var Observable_1 = require("./Observable");
/** An observable object with observable properties */
var OProps = /** @class */ (function (_super) {
    tslib_1.__extends(OProps, _super);
    function OProps(source) {
        return _super.call(this, source) || this;
    }
    OProps.prototype.watch = function (prop) {
        var watched = this.watched || (this.watched = Object.create(null));
        return watched[prop] || (watched[prop] = new OProp(this, prop));
    };
    /** @internal */
    OProps.prototype['_onChange'] = function (change) {
        // When a property has an observable value, it notifies us of changes using
        // itself as the target.
        if (change.target !== this) {
            var currProps = this.get();
            // When a change is from Immer, we must check if its state (or its parent state)
            // is related to us. If so, Immer will update our underlying value for us.
            if (change.state) {
                var _a = change.state, copy = _a.copy, parent = _a.parent;
                if (copy == currProps)
                    return;
                // Check against `parent.base` instead of `parent.copy` because changes
                // are processed from bottom-to-top, which means our underlying value
                // has not been updated yet when we are the parent of this change.
                if (parent && parent.base == currProps)
                    return;
            }
            var target = change.target;
            var nextProps = common_1.shallowCopy(currProps);
            nextProps[target.prop] = change.newValue;
            // Bind `this` to its next value.
            common_1.definePrivate(nextProps, symbols_1.$O, this);
            this._rebind(nextProps);
            freeze_1.freeze(nextProps);
            // Reuse the change object, since only we receive it.
            change.target = this;
            change.prop = target.prop;
            this._onChange(change);
            // Reuse it again for the root change.
            change.prop = null;
            change.oldValue = currProps;
            change.newValue = nextProps;
        }
        // Notify property observers first.
        else if (change.prop !== null) {
            var prop = change.prop, oldValue = change.oldValue, newValue = change.newValue;
            var oldObservable = oldValue && oldValue[symbols_1.$O];
            if (newValue) {
                var observable = Observable_1.getObservable(newValue);
                // Plain objects/arrays are made observable automatically.
                if (!observable && !freeze_1.isFrozen(newValue) && immer_1.isDraftable(newValue)) {
                    observable = bindProps(newValue);
                }
                // Always watch any property that has an observable value.
                if (observable && !oldObservable) {
                    this.watch(prop);
                }
            }
            if (this.watched) {
                var target = this.watched[prop];
                if (target) {
                    commit_1.commit(target, oldValue, newValue, null, false, change.state);
                }
            }
        }
        // Notify root observers.
        _super.prototype._onChange.call(this, change);
    };
    return OProps;
}(Observable_1.Observable));
exports.OProps = OProps;
var OProp = /** @class */ (function (_super) {
    tslib_1.__extends(OProp, _super);
    function OProp(parent, prop) {
        var _this = _super.call(this, function () { return parent.get()[prop]; }) || this;
        _this.parent = parent;
        _this.prop = prop;
        /** Exists when the observed property name has an observable value */
        _this._observedValue = undefined;
        _this._observeValue(_this.get());
        return _this;
    }
    /** @internal */
    OProp.prototype['_onChange'] = function (change) {
        if (change.target === this) {
            this._observeValue(change.newValue);
            _super.prototype._onChange.call(this, change);
        }
        // Changes to `_observedValue` are first handled by our parent.
        else if (change.prop === null) {
            this.parent._onChange(tslib_1.__assign({}, change, { target: this }));
        }
    };
    OProp.prototype.activate = function () {
        // Already active if the value is observable.
        if (!this._observedValue) {
            var watched = this.parent.watched;
            var target = watched[this.prop];
            if (!target) {
                watched[this.prop] = this;
            }
            else if (target !== this) {
                // Never store an OProp longer than you need to.
                throw Error('Cannot observe an old observable');
            }
        }
    };
    OProp.prototype.deactivate = function () {
        // Skip deactivation unless the value is *not* observable.
        // Otherwise, deep observability is prevented.
        if (!this._observedValue) {
            delete this.parent.watched[this.prop];
        }
    };
    OProp.prototype._observeValue = function (value) {
        // Avoid throwing on stale values.
        var observed = value && value[symbols_1.$O];
        if (observed) {
            if (observed !== this._observedValue) {
                if (this._observedValue) {
                    this._observedValue._removeObserver(this);
                }
                if (!this._observers) {
                    this.activate();
                }
                observed._addObserver(this, true);
                this._observedValue = observed;
            }
        }
        else if (this._observedValue) {
            this._observedValue._removeObserver(this);
            this._observedValue = undefined;
            if (!this._observers) {
                this.deactivate();
            }
        }
    };
    return OProp;
}(Observable_1.Observable));
exports.OProp = OProp;
function bindProps(root) {
    if (freeze_1.isFrozen(root)) {
        throw Error('Frozen objects cannot become observable');
    }
    if (immer_1.isDraft(root)) {
        throw Error('Immer drafts cannot become observable');
    }
    if (!immer_1.isDraftable(root)) {
        throw Error('This object has no observable type');
    }
    var rootObservable = new OProps(root);
    common_1.definePrivate(root, symbols_1.$O, rootObservable);
    var observeTree = function (parent, observer) {
        // Only enumerable keys are made observable.
        Object.keys(parent).forEach(function (prop) {
            var value = parent[prop];
            if (common_1.isObject(value)) {
                var observable = Observable_1.getObservable(value);
                if (!observable) {
                    if (freeze_1.isFrozen(value) || !immer_1.isDraftable(value))
                        return;
                    observable = new OProps(value);
                    common_1.definePrivate(value, symbols_1.$O, observable);
                    observeTree(value, observable);
                }
                // To support deep observation, any property with an
                // observable value must be watched.
                observer.watch(prop);
            }
        });
        freeze_1.freeze(parent);
    };
    // The entire state tree is made observable (when possible).
    observeTree(root, rootObservable);
    return rootObservable;
}
exports.bindProps = bindProps;
