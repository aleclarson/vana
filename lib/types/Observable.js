"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("../common");
var debug_1 = require("../debug");
var commit_1 = require("../funcs/commit");
var symbols_1 = require("../symbols");
var nextId = 1;
/** An observable value */
var Observable = /** @class */ (function () {
    function Observable(source) {
        this['_observers'] = undefined;
        this.id = nextId++;
        this._rebind(source);
    }
    /** Try coercing the given value into an Observable instance. */
    Observable.from = function (value) {
        return value instanceof Observable ? value : value ? value[symbols_1.$O] : undefined;
    };
    /** The current value. */
    Observable.prototype.get = function () {
        return this._get ? this._get() : this._value;
    };
    /**
     * When the source of truth is a getter, this method only notifies observers.
     * Otherwise, this method also updates the current value.
     */
    Observable.prototype.set = function (newValue) {
        var oldValue = this.get();
        if (newValue !== oldValue) {
            if (!this._get) {
                this._value = newValue;
            }
            commit_1.commit(this, oldValue, newValue);
        }
    };
    /** Listen for changes. */
    Observable.prototype.tap = function (onUpdate) {
        var _this = this;
        var observer = {
            onUpdate: onUpdate,
            dispose: function () { return _this._removeObserver(observer); },
            _onChange: function (change) { return change.prop !== null || onUpdate(change.newValue); },
        };
        this._addObserver(observer);
        return observer;
    };
    /** Remove all observers and perform subclass cleanup (if necessary) */
    Observable.prototype.dispose = function () {
        if (this._observers) {
            this._observers = undefined;
            this.deactivate();
        }
    };
    /** @internal Notify observers of a change */
    Observable.prototype['_onChange'] = function (change) {
        var observers = this._observers;
        if (debug_1.getDebug(this)) {
            // tslint:disable-next-line
            console.debug('%s(%s) changed: (observers: %O, change: %O)', this.constructor.name, this.id, observers ? observers.length : 0, common_1.shallowCopy(change) // TODO: Deep copy instead
            );
        }
        if (observers) {
            for (var _i = 0, _a = observers.slice(0); _i < _a.length; _i++) {
                var observer = _a[_i];
                observer._onChange(change);
            }
        }
    };
    /** @internal Returns false when the given value is outdated */
    Observable.prototype['_isCurrent'] = function (value) {
        return value === this.get();
    };
    /** @internal Swap out the source of truth for this observable. */
    Observable.prototype['_rebind'] = function (source) {
        this._get = typeof source == 'function' ? source : undefined;
        this._value = this._get ? undefined : source;
    };
    /** @internal */
    Observable.prototype['_addObserver'] = function (observer, prepend) {
        if (this._observers) {
            this._observers[prepend ? 'unshift' : 'push'](observer);
        }
        else {
            this.activate();
            this._observers = [observer];
        }
        return observer;
    };
    /** @internal */
    Observable.prototype['_removeObserver'] = function (observer) {
        var observers = this._observers;
        if (observers) {
            var index = observers.indexOf(observer);
            if (index < 0)
                return;
            if (observers.length > 1) {
                observers.splice(index, 1);
            }
            else {
                this._observers = undefined;
                this.deactivate();
            }
        }
    };
    /** Called when an Observable goes from 0 observers to 1+ */
    Observable.prototype.activate = function () { };
    /** Called when an Observable goes from 1+ observers to 0 */
    Observable.prototype.deactivate = function () { };
    return Observable;
}());
exports.Observable = Observable;
/** Returns true if the given value has been passed to `o()` and is not stale. */
function isObservable(value) {
    if (value) {
        var observable = value[symbols_1.$O];
        if (observable) {
            return observable._isCurrent(value);
        }
    }
    return false;
}
exports.isObservable = isObservable;
/** Throws an error when the given object is an old revision */
function getObservable(value) {
    var observable = value[symbols_1.$O];
    if (observable) {
        if (!observable._isCurrent(value)) {
            throw Error('Outdated values cannot be observed');
        }
        return observable;
    }
}
exports.getObservable = getObservable;
/** @internal */
function getObservers(value) {
    var observable = Observable.from(value);
    if (observable) {
        return observable._observers;
    }
    throw Error('Not observable');
}
exports.getObservers = getObservers;
