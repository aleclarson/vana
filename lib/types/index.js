"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Observable_1 = require("./Observable");
exports.Observable = Observable_1.Observable;
var OBinding_1 = require("./OBinding");
exports.OBinding = OBinding_1.OBinding;
var OPath_1 = require("./OPath");
exports.OPath = OPath_1.OPath;
var OPromise_1 = require("./OPromise");
exports.OPromise = OPromise_1.OPromise;
var OProps_1 = require("./OProps");
exports.OProps = OProps_1.OProps;
var OTuple_1 = require("./OTuple");
exports.OTuple = OTuple_1.OTuple;
