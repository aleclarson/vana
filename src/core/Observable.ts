import { getDebug } from '../debug'
import { $O, IDisposable, isFunction, shallowCopy } from '../shared'
import { Change, IChangeTarget } from './Change'
import { commit } from './commit'

/**
 * Observers are disposable "change targets".
 *
 * Use `IChangeTarget<T>` if your observer is not disposable.
 */
export interface IObserver<T = any> extends IChangeTarget<T>, IDisposable {}

/** The source of truth for an observable */
type Source<T = any> = (() => T) | T

let nextId = 1

/** An observable value */
export class Observable<T = any> implements IObserver<T> {
  ['_observers']?: Array<IChangeTarget<T>> = undefined
  protected _get?: () => T
  protected _value?: T

  readonly id: number

  constructor(source: Source<T>) {
    this.id = nextId++
    this._rebind(source)
  }

  /** Try coercing the given value into an Observable instance. */
  static from<T = any>(value: any): Observable<T> | undefined {
    return value instanceof Observable ? value : value ? value[$O] : undefined
  }

  /** The current value. */
  get(): T {
    return this._get ? this._get() : this._value!
  }

  /**
   * When the source of truth is a getter, this method only notifies observers.
   * Otherwise, this method also updates the current value.
   */
  set(newValue: T, force?: boolean) {
    let oldValue = this.get()
    if (force || newValue !== oldValue) {
      if (!this._get) {
        this._value = newValue
      }
      commit(this, oldValue, newValue)
    }
  }

  /** Listen for changes. */
  tap(onUpdate: (value: T) => void) {
    type ITapper = IObserver<T> & {
      onUpdate: typeof onUpdate
    }
    let observer: ITapper = {
      onUpdate,
      dispose: () => this._removeObserver(observer),
      _onChange: change =>
        change.prop === null && onUpdate(change.newValue as any),
    }
    this._addObserver(observer)
    return observer
  }

  /** Remove all observers and perform subclass cleanup (if necessary) */
  dispose() {
    if (this._observers) {
      this._observers = undefined
      this.deactivate()
    }
  }

  /** @internal Notify observers of a change */
  ['_onChange'](change: Change<T>) {
    let observers = this._observers
    if (getDebug(this)) {
      // tslint:disable-next-line
      console.debug(
        '%s(%s) changed: (observers: %O, change: %O)',
        this.constructor.name,
        this.id,
        observers ? observers.length : 0,
        shallowCopy(change) // TODO: Deep copy instead
      )
    }
    if (observers) {
      for (let observer of observers.slice(0)) {
        observer._onChange(change)
      }
    }
  }

  /** @internal Returns false when the given value is outdated */
  ['_isCurrent'](value: any) {
    return value === this.get()
  }

  /** @internal Swap out the source of truth for this observable. */
  ['_rebind'](source: Source<T>) {
    this._get = isFunction(source) ? source : undefined
    this._value = this._get ? undefined : (source as any)
  }

  /** @internal */
  ['_addObserver'](observer: IChangeTarget<T>, prepend?: boolean) {
    if (this._observers) {
      this._observers[prepend ? 'unshift' : 'push'](observer)
    } else {
      this.activate()
      this._observers = [observer]
    }
    return observer
  }

  /** @internal */
  ['_removeObserver'](observer: IChangeTarget<T>) {
    let observers = this._observers
    if (observers) {
      let index = observers.indexOf(observer)
      if (index < 0) return
      if (observers.length > 1) {
        observers.splice(index, 1)
      } else {
        this._observers = undefined
        this.deactivate()
      }
    }
  }

  /** Called when an Observable goes from 0 observers to 1+ */
  protected activate() {}

  /** Called when an Observable goes from 1+ observers to 0 */
  protected deactivate() {}
}

/** Returns true if the given value has been passed to `o()` and is not stale. */
export function isObservable(value: any): value is object {
  if (value) {
    let observable: Observable = value[$O]
    if (observable) {
      return observable._isCurrent(value)
    }
  }
  return false
}

/** Throws an error when the given object is an old revision */
export function getObservable<T = any>(value: object) {
  let observable: Observable<T> = value[$O]
  if (observable) {
    if (!observable._isCurrent(value)) {
      throw Error('Outdated values cannot be observed')
    }
    return observable
  }
}

/** @internal */
export function getObservers(value: object) {
  let observable = Observable.from(value)
  if (observable) {
    return observable._observers
  }
  throw Error('Not observable')
}
