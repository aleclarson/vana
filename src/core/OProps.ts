// tslint:disable:variable-name
import { isDraft, isDraftable } from 'immer'
import { $O, definePrivate, Dictionary, isObject, shallowCopy } from '../shared'
import { Change } from './Change'
import { commit } from './commit'
import { freeze, isFrozen } from './freeze'
import { getObservable, Observable } from './Observable'

/** An observable object with observable properties */
export class OProps<T extends Dictionary<any> = any> extends Observable<T> {
  watched?: Dictionary<OProp<T>>

  constructor(source: Exclude<T, Function>) {
    super(source)
  }

  watch<P extends string | number>(prop: P): OProp<T, P> {
    const watched = this.watched || (this.watched = Object.create(null))
    return watched[prop] || (watched[prop] = new OProp<T, P>(this, prop))
  }

  /** @internal */
  ['_onChange'](change: Change<T>) {
    // The target is either this or an OProp object whose value is observable.
    if (change.target !== this) {
      const currProps = this.get()

      // When a change is from Immer, we must check if its state (or its parent state)
      // is related to us. If so, Immer will update our underlying value for us.
      if (change.state) {
        const { copy, parent } = change.state
        if (copy == currProps) return
        // Check against `parent.base` instead of `parent.copy` because changes
        // are processed from bottom-to-top, which means our underlying value
        // has not been updated yet when we are the parent of this change.
        if (parent && parent.base == currProps) return
      }

      const target = change.target as OProp<T>
      const nextProps = shallowCopy(currProps)
      nextProps[target.prop] = change.newValue

      // Bind `this` to its next value.
      definePrivate(nextProps, $O, this)
      this._rebind(nextProps)
      freeze(nextProps)

      // Reuse the change object, since only we receive it.
      change.target = this
      change.prop = target.prop
      this._onChange(change)

      // Reuse it again for the root change.
      change.prop = null
      change.oldValue = currProps
      change.newValue = nextProps
    }

    // Property changes are handled before notifying root observers.
    else if (change.prop !== null) {
      const { prop, oldValue, newValue } = change

      // Ensure the new value is observable if it can be.
      const oldObservable = oldValue && oldValue[$O]
      if (newValue) {
        let observable = getObservable(newValue)

        // Plain objects/arrays are made observable automatically.
        if (!observable && !isFrozen(newValue) && isDraftable(newValue)) {
          observable = bindProps(newValue)
        }

        // Always watch any property that has an observable value.
        if (observable && !oldObservable) {
          this.watch(prop)
        }
      }

      // Notify any property observers.
      if (this.watched) {
        const target = this.watched[prop]
        if (target) {
          commit(target, oldValue, newValue, null, false, change.state)
        }
      }
    }

    // Notify any root observers.
    super._onChange(change)
  }
}

export class OProp<
  T extends Dictionary<any> = any,
  P extends string | number = any
> extends Observable<T[P]> {
  /** Exists when the observed property name has an observable value */
  private _observedValue?: Observable<T[P]> = undefined

  constructor(readonly parent: OProps<T>, readonly prop: P) {
    super(() => parent.get()[prop])
    this._observeValue(this.get())
  }

  /** @internal */
  ['_onChange'](change: Change<T[P]>) {
    if (change.target === this) {
      this._observeValue(change.newValue)
      super._onChange(change)
    }
    // Changes to `_observedValue` are first handled by our parent.
    else if (change.prop === null) {
      this.parent._onChange({ ...change, target: this })
    }
  }

  protected activate() {
    // Already active if the value is observable.
    if (!this._observedValue) {
      const watched = this.parent.watched!
      const target = watched[this.prop]
      if (!target) {
        watched[this.prop] = this
      } else if (target !== this) {
        // Never store an OProp longer than you need to.
        throw Error('Cannot observe an old observable')
      }
    }
  }

  protected deactivate() {
    // Skip deactivation unless the value is *not* observable.
    // Otherwise, deep observability is prevented.
    if (!this._observedValue) {
      delete this.parent.watched![this.prop]
    }
  }

  private _observeValue(value: T[P]) {
    // Avoid throwing on stale values.
    const observed = value && value[$O]
    if (observed) {
      if (observed !== this._observedValue) {
        if (this._observedValue) {
          this._observedValue._removeObserver(this)
        }
        if (!this._observers) {
          this.activate()
        }
        observed._addObserver(this, true)
        this._observedValue = observed
      }
    } else if (this._observedValue) {
      this._observedValue._removeObserver(this)
      this._observedValue = undefined

      if (!this._observers) {
        this.deactivate()
      }
    }
  }
}

export function bindProps<T extends object>(root: Exclude<T, Function>) {
  if (isFrozen(root)) {
    throw Error('Frozen objects cannot become observable')
  }
  if (isDraft(root)) {
    throw Error('Immer drafts cannot become observable')
  }
  if (!isDraftable(root)) {
    throw Error('This object has no observable type')
  }

  const rootObservable = new OProps(root)
  definePrivate(root, $O, rootObservable)

  const observeTree = (parent: object, observer: OProps) => {
    // Only enumerable keys are made observable.
    Object.keys(parent).forEach(prop => {
      const value = parent[prop]
      if (isObject(value)) {
        let observable = getObservable(value) as OProps
        if (!observable) {
          if (isFrozen(value) || !isDraftable(value)) return
          observable = new OProps(value)
          definePrivate(value, $O, observable)
          observeTree(value, observable)
        }
        // To support deep observation, any property with an
        // observable value must be watched.
        observer.watch(prop)
      }
    })
    freeze(parent)
  }

  // The entire state tree is made observable (when possible).
  observeTree(root, rootObservable)
  return rootObservable
}
