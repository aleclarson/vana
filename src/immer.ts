import { Draft, Immer, IProduce } from 'immer'
import { definePrivate, isArray } from './common'
import { commit } from './funcs/commit'
import { __$observable, Observable } from './types/Observable'

/** Function that modifies a draft */
export type Recipe<
  T extends object = any,
  Args extends any[] = any[],
  Result = any
> = (draft: Draft<T>, ...args: Args) => Result

export const immer = new Immer({
  onAssign(state, prop: any) {
    state.assigned[prop] = true
  },
  onDelete(state, prop: any) {
    state.assigned[prop] = false
  },
  onCopy(state) {
    let { base } = state
    let target: Observable = base && base[__$observable]
    if (target) {
      let { copy, assigned } = state
      if (isArray(base)) {
        // Non-arrays have their __$observable property copied over by Immer.
        definePrivate(copy, __$observable, target)
      }
      target._rebind(copy)
      for (let prop in assigned) {
        commit(target, base[prop], copy[prop], prop, !assigned[prop], state)
      }
      commit(target, base, copy, null, false, state)
    }
  },
})

// tslint:disable:no-default-export
export default immer.produce

export const produce: IProduce = immer.produce
export const setAutoFreeze = immer.setAutoFreeze.bind(immer)
export const setUseProxies = immer.setUseProxies.bind(immer)
