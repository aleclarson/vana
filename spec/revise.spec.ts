import { original } from 'immer'
import { isObservable, o, revise, tap } from '../src'
import { getObservable, getObservers } from '../src/types/Observable'
import { OProps } from '../src/types/OProps'

describe('revise()', () => {
  describe('when the 2nd argument is a plain object', () => {
    it('cannot revise a stale object', () => {
      let base = o({ a: 1 })
      revise(base, { a: 2 })
      expect(() => {
        revise(base, { a: 3 })
      }).toThrowErrorMatchingSnapshot()
    })

    it('is a no-op if no values changed', () => {
      let base = o({ a: 1 })
      expect(revise(base, {})).toBe(base)
      expect(revise(base, { a: 1 })).toBe(base)
    })

    it('never mutates the base object', () => {
      let base = o({ a: 1 })
      let copy = revise(base, { a: 2 })
      expect(copy.a).toBe(2)
      expect(base.a).toBe(1)
    })

    it('binds the Observable to the copy', () => {
      let base = o({ a: 1 })
      let baseObservable = getObservable(base)!

      let copy = revise(base, { a: 2 })
      expect(baseObservable).toBe(getObservable(copy))
      expect(baseObservable.get()).toBe(copy)
    })

    it('always does a shallow copy', () => {
      let base = { a: { b: 1 }, c: 1 }
      let changes = { a: { b: 2 } }
      let copy = revise(base, changes)
      expect(copy.a).toBe(changes.a)
      expect(copy).toEqual({ ...base, ...changes })
    })

    it('throws when the base object has a getter', () => {
      let base = {}
      Object.defineProperty(base, 'a', {
        get: () => 1,
        enumerable: true,
      })

      expect(() => {
        revise(base, { a: 2 })
      }).toThrowErrorMatchingSnapshot()
    })

    it('preserves non-enumerable properties', () => {
      let base = { a: 1 } as { a: number; b: number }
      Object.defineProperty(base, 'b', {
        value: 1,
      })

      let copy = revise(base, { a: 2 })
      expect(copy.a).toBe(2)
      expect(copy.b).toBe(1)

      copy = revise(copy, { b: 2 })
      expect(copy.a).toBe(2)
      expect(copy.b).toBe(2)
    })

    it('preserves the prototype of null objects', () => {
      let base = Object.create(null)
      let copy = revise(base, { a: 1 })
      expect(Object.getPrototypeOf(copy)).toBe(null)
      expect(copy.a).toBe(1)
    })

    it('throws when the base object is a class instance', () => {
      class Foo {}
      let base = new Foo()
      expect(() => {
        revise(base, { a: 1 })
      }).toThrowErrorMatchingSnapshot()
    })

    describe('when a new value is a plain object', () => {
      it('makes the value deeply observable', () => {
        let base: { a: any } = o({ a: 1 })
        let copy = revise(base, {
          a: {
            b: {},
            c: Object.freeze({}), // Frozen objects are skipped.
          },
        })
        expect(isObservable(copy.a)).toBeTruthy()
        expect(isObservable(copy.a.b)).toBeTruthy()
        expect(isObservable(copy.a.c)).toBeFalsy()
      })
    })

    // TODO: it('works inside a producer', () => {})
  })

  describe('when the 2nd argument is a recipe function', () => {
    it('can delete a property', () => {
      let base = o({ a: 1 })
      let spy = jest.fn()
      tap(base, spy)

      let copy = revise(base, draft => {
        delete draft.a
      })
      expect(spy).toBeCalledWith(copy)
    })

    describe('when changes are made', () => {
      it('returns an observable copy', () => {
        type Base = { a: number; b?: number }
        let base: Base = o({ a: 0 })
        let copy = revise(base, draft => {
          draft.a = 1
          draft.b = 1
        })

        expect(copy).not.toBe(base)
        expect(copy).toEqual({ a: 1, b: 1 })

        expect(isObservable(copy)).toBeTruthy()
        expect(isObservable(base)).toBeFalsy()
      })
    })

    describe('when nothing changes', () => {
      it('returns the original object', () => {
        let base = o({ a: 0 })
        let copy = revise(base, () => {})
        expect(copy).toBe(base)
      })
    })

    it('passes a mutable draft to the recipe function', () => {
      let base = o({ a: 0 })
      revise(base, draft => {
        expect(draft).not.toBe(base)
        draft.a = 1
        expect(draft.a).toBe(1)
      })
    })

    it('can pass extra arguments to the recipe function', () => {
      let obj = o({ a: 0 })
      let args = [1, 2, 3, 4, 5, 6]
      revise(
        obj,
        (draft: any, ...rest: any[]) => {
          expect(original(draft)).toBe(obj)
          expect(rest).toEqual(args)
        },
        ...args
      )
    })

    it('works with non-observable objects', () => {
      type Base = { a: number; b: number; c?: number }
      let base: Base = { a: 0, b: 0 }
      let changes = { a: 1, c: 1 }
      let copy = revise(base, changes)
      expect(base).not.toBe(copy)
      expect(copy).toEqual({ ...base, ...changes })
      expect(revise(copy, { a: 1 })).toBe(copy)
    })

    it('supports replacing an observable value with another', () => {
      let a1 = o({})
      let base = o({ a: a1 })
      let baseObservable = getObservable(base) as OProps
      let propObservable = baseObservable.watched!.get('a')!

      let a2 = o({})
      revise(base, draft => {
        draft.a = a2
      })
      expect(getObservers(a1)).toBeUndefined()
      expect(getObservers(a2)).toEqual([propObservable])
    })
  })
})
