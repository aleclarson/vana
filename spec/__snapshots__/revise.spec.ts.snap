// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`revise() when the 2nd argument is a plain object cannot revise a stale object 1`] = `"Outdated values cannot be observed"`;

exports[`revise() when the 2nd argument is a plain object throws when the base object has a getter 1`] = `"Cannot mutate a computed property"`;

exports[`revise() when the 2nd argument is a plain object throws when the base object is a class instance 1`] = `"Expected a plain object"`;
